tests/testsuite/registry.rs

```diff
@@ -12,8 +14,52 @@ use std::io::{BufRead, BufReader, Write};
 use std::path::Path;
 use std::process::Stdio;
 
+fn cargo_http(p: &Project, s: &str) -> Execs {
+    let mut e = p.cargo(s);
+    e.arg("-Zhttp-registry").masquerade_as_nightly_cargo();
+    e
+}
+
+fn cargo_stable(p: &Project, s: &str) -> Execs {
+    p.cargo(s)
+}
+
+fn setup_http() -> RegistryServer {
+    let server = serve_registry(registry_path());
+    configure_source_replacement_for_http(&server.addr().to_string());
+    server
+}
+
+fn configure_source_replacement_for_http(addr: &str) {
+    let root = paths::root();
+    t!(fs::create_dir(&root.join(".cargo")));
+    t!(fs::write(
+        root.join(".cargo/config"),
+        format!(
+            "
+            [source.crates-io]
+            replace-with = 'dummy-registry'
+
+            [source.dummy-registry]
+            registry = 'sparse+http://{}'
+        ",
+            addr
+        )
+    ));
+}
+
 #[cargo_test]
-fn simple() {
+fn simple_http() {
+    let _server = setup_http();
+    simple(cargo_http);
+}
+
+#[cargo_test]
+fn simple_git() {
+    simple(cargo_stable);
+}
+
+fn simple(cargo: fn(&Project, &str) -> Execs) {
```

Comment:
The repeated boilerplate in these tests is acceptable. And far better than it was. And still kind of unfortunate. I don't know enough about proc macros, I wonder if we could have `#[cargo_test(HTTP)] fn simple` generate all three functions?

Definitely not something we have to fix now, but something I want to remember to discuss.